moped-editor/src/views/projects/projectView/ProjectComponents/utils/useCreateComponent.js

```diff
@@ -0,0 +1,230 @@
+
+      const draftFeaturesToKeep = state.draftComponent.features.filter(
+        (feature) => {
+          if (isDrawnFeature(feature)) {
+            return !featureIdsToDelete.includes(getDrawId(feature));
+          } else {
+            return true;
+          }
+        }
+      );
+
+      const draftComponentWithDeletes = {
+        ...state.draftComponent,
+        features: [...draftFeaturesToKeep],
+      };
+
+      return { ...state, draftComponent: draftComponentWithDeletes };
+    default:
+      throw Error(`Unknown action. ${action.type}`);
+  }
+};
+
+export const useCreateComponent = ({
+  projectId,
+  setClickedComponent,
+  setLinkMode,
+  refetchProjectComponents,
+}) => {
+  const [createState, createDispatch] = useReducer(createReducer, {
+    isCreatingComponent: false,
+    showCreateDialog: false,
+    draftComponent: null,
+  });
+
+  const [addProjectComponent] = useMutation(ADD_PROJECT_COMPONENT);
+
+  const onStartCreatingComponent = () => {
+    createDispatch({ type: "start_create" });
+    setClickedComponent(null);
+  };
+
+  const onCancelComponentCreate = () => {
+    createDispatch({ type: "cancel_create" });
+    setLinkMode(null);
+  };
+
+  const onSaveDraftComponent = () => {
+    /* Start data preparation */
+    const {
+      component_id,
+      description,
+      moped_subcomponents,
+      component_name,
+      internal_table,
+      features,
+    } = createState.draftComponent;
+
+    const subcomponentsArray = moped_subcomponents
+      ? moped_subcomponents.map((subcomponent) => ({
+          subcomponent_id: subcomponent.value,
+        }))
+      : [];
+
+    const featureTable = internal_table;
+
+    const featuresToInsert = [];
+    const drawnLinesToInsert = [];
+    const drawnPointsToInsert = [];
+
+    const drawnFeatures = features.filter((feature) =>
+      Boolean(getDrawId(feature))
+    );
+    const selectedFeatures = features.filter(
+      (feature) => !Boolean(getDrawId(feature))
+    );
+
+    if (featureTable === "feature_street_segments") {
+      makeLineStringFeatureInsertionData(
+        featureTable,
+        selectedFeatures,
+        featuresToInsert
+      );
+      makeDrawnLinesInsertionData(drawnFeatures, drawnLinesToInsert);
+    } else if (
+      featureTable === "feature_intersections" ||
+      featureTable === "feature_signals"
+    ) {
+      makePointFeatureInsertionData(
+        featureTable,
+        selectedFeatures,
+        featuresToInsert
+      );
+      makeDrawnPointsInsertionData(drawnFeatures, drawnPointsToInsert);
+    }
+
+    const newComponentData = {
+      description,
+      component_id,
+      name: component_name,
+      project_id: projectId,
+      moped_proj_components_subcomponents: {
+        data: subcomponentsArray,
+      },
+      [featureTable]: {
+        data: featuresToInsert,
+      },
+      feature_drawn_lines: { data: drawnLinesToInsert },
+      feature_drawn_points: { data: drawnPointsToInsert },
+    };
```

Comment:
I'll look into this as well. I want to find a better way here to because, once again, this only gets more complex with editing drawn component features.

Reply:
Going to take another swing at this in the next PR. Breaking down some of the large blocks is at top of mind and that fits this same goal. I like where you are going with this.